### PR TITLE
ConfidentalClientApplicationBuilder WithSendX5C

### PR DESCRIPTION
**Pull request by hpsin**
Proposes a solution for #2804

**Changes proposed in this request**
Add WithSendX5C to ConfidentialClientApplicationBuilder. 
Makes the existing SendX5C channeled throughout confidential clients nullable, to be overwritten at request time by the default-null SendX5C within the ConfidentialClient. 
Marks the per-request SendX5C Builder methods as Obsolete, so that devs know to switch to per-app X5C. 

**Testing**
Inline updates to the existing unit tests, to switch away from the per-request model and into the per-app model.  This should validate that the claim is appearing as expected as a result of this change. 

**Performance impact**
No perf impact expected.

### CODE REVIEW

**Comment by bgavrilMS** [comment]
Very cool @hpsin

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByRefreshTokenParameterBuilder.cs
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByRefreshTokenParameterBuilder.cs:74]: If the user sets the x5c flag here to false, and is already set in the config as true, I think we should throw an exception. The config object can be access by: 

`this.ServiceBundle.Config`

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:145]: Does this work? WithSendX5C should not increment this counter, which simply counts how many times APIs such as `WithSecret` `WithCertificate` etc. were used. It's used simply to throw an exception when 0 or >1 secrets are configured,.

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
LGTM, there is one small comment that needs to be resolved.

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:108]: `WithClientClaims` also needs to be modified to take in this flag.

**Comment by bgavrilMS** [comment]
@jmprieur @jennyf19 - please have a look at this PR / public API surface.

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs
Comment (hpsin) [reply src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:108]: So, I wasn't sure if WithClientClaims could itself support injecting the x5c claim. I was worried that we might collide. If it only injects claims into the body, which makes sense, then yeah, for sure.

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs
Comment (hpsin) [reply src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:145]: My thinking was that this would cause Build to throw if you had set WithSendX5C and not set a certificate. That would be expected, because you can only use x5c with a cert (and not with a secret or provided assertion).

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:145]: Yeah, I agree that if you set sendX5C but don't use WithCertificate, we should throw. I don't think we do this now, and there is no need to do it as part of this PR.

But when we throw exceptions, we need to be very explicit in describing the error message. In this case what would happen is that if the developer adds 

`app.WithCertificate(cert, true)` then that counter would be 2 and MSAL would throw an exception thinking "oh, you've used WithCertificate and WithSecret at the same time, this is not supported".

Comment (hpsin) [reply src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:145]: Oh, I totally misread this to be a tracker on cert use, not credentials attached. Yep, will remove. Thanks!

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByAuthorizationCodeParameterBuilder.cs
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByAuthorizationCodeParameterBuilder.cs:88]: I don't think we should obsolete the one on the AT calls. Both should work. We can change documentation to reflect the one on the cca object, but let's not break people.

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs:14]: >? [](http://example.com/codeflow?start=19&length=1)

i don't think this will work in MSAL, they don't support the nullable reference types introduced in c# 8.

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:137]: i don't think we need this one. It has to go with the cert or signed assertion, so makes sense that it's an additional parameter there, so no need to have it separate. I think it would lead to confusion and people setting it twice.

File: src/client/Microsoft.Identity.Client/AppConfig/IAppConfig.cs
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/AppConfig/IAppConfig.cs:131]: if we go w/the parameter option, this is not needed.

File: src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs:200]: >credential [](http://example.com/codeflow?start=60&length=10)

null check on `credential`?

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByAuthorizationCodeParameterBuilder.cs
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByAuthorizationCodeParameterBuilder.cs:88]: I agree with Jenny that this will be an additional method and need not obsolete

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs:14]: Help me understand why nullable bool is required and not just bool?

**Review by jennyf19 — CHANGES_REQUESTED** [review]
:clock1:

**Review by SameerK-MSFT — CHANGES_REQUESTED** [review]

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs:14]: @jennyf19 - nullable types were introduced in very early C# version, we do support them in MSAL. However, nullable *reference* types are not supported because of C# 8 - https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references

File: src/client/Microsoft.Identity.Client/AppConfig/IAppConfig.cs
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/AppConfig/IAppConfig.cs:131]: This comment reminds me that it would be nice to support x5c as an option.

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs:14]: That's what I thought I wrote. `nullable reference types`

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs
Comment (hpsin) [reply src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:113]: Generally agreed. Thankfully setting it twice won't do anything bad. The only reason I went this way was to snap with how the other SDKs do it (e.g. [Java](https://www.javadoc.io/doc/com.microsoft.azure/msal4j/latest/com/microsoft/aad/msal4j/ConfidentialClientApplication.Builder.html)) - happy to remove.  Other consideration was that since SNI is still largely 1p only, I wasn't sure about polluting the public namespace so visibly with something they shouldn't use. WDYT - remove it?

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByAuthorizationCodeParameterBuilder.cs
Comment (hpsin) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByAuthorizationCodeParameterBuilder.cs:88]: Right, good for overrides if such a thing is needed.  I was expecting that this would be a warning. not an error, so not break people.  If the library setup to treat this as an error despite passing false into the Obsolete?

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs
Comment (hpsin) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs:14]: Ah, darn.  This makes things much trickier - I'm worried that without being nullable it will default to false on a per-request basis, and overwrite the cca field that should be the primary driver of the value. 

Goal here: If the request params explicitly has SendX5C set, use that, otherwise fall back to the default-false cca object. Currently I'm using the nullable as a way to signal it's been explicitly set.  Should I add a new field instead, that's set during the per-request call, something like `bool overrideCcaSendX5C` in `IAcquireTokenParameters`?

File: src/client/Microsoft.Identity.Client/AppConfig/IAppConfig.cs
Comment (hpsin) [reply src/client/Microsoft.Identity.Client/AppConfig/IAppConfig.cs:131]: Ah, what's the parameter option? Embedding it inside the credentialWrapper instead of inside the IAppConfig?

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByAuthorizationCodeParameterBuilder.cs
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByAuthorizationCodeParameterBuilder.cs:88]: This will emit a warning, not an error. We do this from time to time. It will not break people, but it will annoy them. 

This will warn ALL developers using `sendx5c` (1st parties only I guess) to upgrade. Some projects (e.g. Kusto) are very strict and have a 0 warnings policy. Others won't care. 

If we think it's wrong to set sendX5C at request level, let's deprecate it. But we must make sure to update our 1st party docs and that the obsolete message is very clear. 

@henrik-me - thoughts?

Aside: When we release MSAL 5 (we keep saying this for the past 3 years), we will delete all the deprecated APIs.

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs:14]: Sorry, I mean that nullable **value** types was introduced since c# 2.0. Nullable reference types was introduced in C# 8.0

image

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:113]: I would do one or the other, but not both. I guess Jean-Marc would need to decide. What happens if a 3rd party tries to use it? also it's false by default, so most won't even notice it.

Comment (hpsin) [reply src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:113]: 3ps that try to use this will see a perf impact, but won't break (it doesn't help or harm them beyond sending extraneous bits over the wire).  Classic AAD behaviour would be to ignore the x5c claim if the cert thumbprint exists already in the app/SP registration, because while the developer is doing something "wrong" they're not doing anything illegal, so we'll let them do it.

Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:113]: It's kind of a non-optional parameter for first party, so i think it goes outside our normal "let's not pollute". 
@jmprieur to decide.

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs:14]: @hpsin I think it would be better to modify the logic internally to choose the proper value based on the parameters set rather than adding additional api here.

**Review by trwalke — CHANGES_REQUESTED** [review]
Remove nullable reference type and update internal logic to handle sendX5c when set from various locations in MSAL

**Comment by trwalke** [comment]
Also, Since this is a fork, the automation did not run. We should probably recreate this PR with a branch on our main repo and allow the automation to run.

**Comment by jennyf19** [comment]
we can make a branch off of master and merge into that branch once @hpsin is done and then we can check the tests and make any final changes before merging into master.

---
In reply to: [891407708](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2805#issuecomment-891407708) [](http://example.com/codeflow?ancestors=891407708,891407708)

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs
Comment (hpsin) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenByAuthorizationCodeParameters.cs:14]: Can you clarify exactly what you mean by modify the logic internally? I don't thing I have the same name references for things that your team does :(  The specific line you're commenting on makes the notion of setting X5C per-request nullable so that we can perform internal logic to figure out what to do, rather than adding fields to the request parameters object. We can't do this without an API in general, as the developer needs to have some way of specifying whether we do this.

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs
Comment (hpsin) [inline src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs:19]: The amount of repetition here makes me wonder if there should be an IConfidentialClientRequestParameters.

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByAuthorizationCodeParameterBuilder.cs
Comment (henrik-me) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenByAuthorizationCodeParameterBuilder.cs]: +1 to Jenny: Let's not remove per request override and thus I see no reason to introduce a warning here. If anything an analyzer needs to be written to ensure recommended practices are followed.

(btw Java sends the X5C by default but allows to disable sending it)

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs
Comment (henrik-me) [reply src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:113]: Could be an option to change the default of sending the X5C to true

**Review by jennyf19 — APPROVED** [review]
:shipit:

**Review by trwalke — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs:19]: There is, it's called `AbstractClientAppBaseAcquireTokenParameterBuilder`

**Review by trwalke — APPROVED** [review]
Approving so this can be recreated on our repo and finalized

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:113]: It is server to server communication (i.e. high bandwidth). So it impact may not be as much. It is a trade of between reliability and perf. The perf impact will be smaller compared to the gain. - My 2 cents

**Pull request merged**
